src/CoreWCF.Http/tests/Helpers/IntegrationTest.cs

```diff
@@ -0,0 +1,53 @@
+﻿using System;
+using System.IO;
+using System.Net.Http;
+using Helpers;
+using Microsoft.AspNetCore.Hosting;
+using Microsoft.AspNetCore.Hosting.Server.Features;
+using Microsoft.AspNetCore.Http.Features;
+using Microsoft.AspNetCore.Mvc.Testing;
+using Microsoft.AspNetCore.TestHost;
+
+namespace CoreWCF.Http.Tests.Helpers
+{
+    /// <summary>
+    /// Enables in-memory integration testing for CoreWCF (outside-in testing via <see cref="HttpClient"/>).
+    ///
+    /// Use these tests to exercise the entire HTTP stack, rather than create in-process ServiceModel channels.
+    /// 
+    /// <see href="https://docs.microsoft.com/en-us/aspnet/core/test/integration-tests?view=aspnetcore-3.1"/>
+    /// <seealso href="https://docs.microsoft.com/en-us/aspnet/core/test/integration-tests?view=aspnetcore-2.1"/>
+    /// </summary>
+    /// <typeparam name="TStartup"></typeparam>
+    public class IntegrationTest<TStartup> : WebApplicationFactory<TStartup> where TStartup : class
+    {
+        protected override TestServer CreateServer(IWebHostBuilder builder)
+        {       
+            var addresses = new ServerAddressesFeature();
+            var features = new FeatureCollection();
+            features.Set<IServerAddressesFeature>(addresses); 
+
+            var server = new TestServer(builder, features);
+#if NETCOREAPP3_1
+            server.AllowSynchronousIO = true;
+#endif
+            return server;
+        }
+
+        protected override IWebHostBuilder CreateWebHostBuilder()
+        {
+            SetSelfHostedContentRoot();
+
+            return ServiceHelper.CreateWebHostBuilder<TStartup>();
+        }
+
+        private static void SetSelfHostedContentRoot()
+        {
+            var contentRoot = Directory.GetCurrentDirectory();
+            var assemblyName = typeof(IntegrationTest<TStartup>).Assembly.GetName().Name;
+            var settingSuffix = assemblyName.ToUpperInvariant().Replace(".", "_");
+            var settingName = $"ASPNETCORE_TEST_CONTENTROOT_{settingSuffix}";
```

Comment:
Again, the content root stuff is only needed for MVC razor pages. Does the test server need it too?

Reply:
The test server does need it. You can see that it crashes if you remove this code. It's only there to route around the `TestServer` check for it, because it has an unguarded exception.